bika/lims/browser/fields/aranalysesfield.py

```diff
@@ -99,158 +122,226 @@ def get(self, instance, **kwargs):
+        # Service UIDs
+        service_uids = map(api.get_uid, services)
+
+        # Modify existing AR specs with new form values of selected analyses.
+        self._update_specs(instance, specs)
+
+        for service in services:
             keyword = service.getKeyword()
 
-            # analysis->InterimFields
-            calc = service.getCalculation()
-            interim_fields = calc and list(calc.getInterimFields()) or []
-
-            # override defaults from service->InterimFields
-            service_interims = service.getInterimFields()
-            sif = dict([(x['keyword'], x.get('value', ''))
-                        for x in service_interims])
-            for i, i_f in enumerate(interim_fields):
-                if i_f['keyword'] in sif:
-                    interim_fields[i]['value'] = sif[i_f['keyword']]
-                    service_interims = [x for x in service_interims
-                                        if x['keyword'] != i_f['keyword']]
-            # Add remaining service interims to the analysis
-            for v in service_interims:
-                interim_fields.append(v)
-
-            # create the analysis if it doesn't exist
+            # Create the Analysis if it doesn't exist
             if shasattr(instance, keyword):
                 analysis = instance._getOb(keyword)
             else:
                 analysis = create_analysis(instance, service)
                 new_analyses.append(analysis)
-            for i, r in enumerate(rr):
-                if r['keyword'] == analysis.getKeyword():
-                    r['uid'] = analysis.UID()
+
+            # Update the interim fields
+            self._update_interims(analysis, service)
+
+            # Set the price of the Analysis
+            self._update_price(analysis, service, prices)
 
         # delete analyses
         delete_ids = []
         for analysis in instance.objectValues('Analysis'):
             service_uid = analysis.getServiceUID()
-            if service_uid not in service_uids:
-                # If it is verified or published, don't delete it.
-                state = workflow.getInfoFor(analysis, 'review_state')
-                if state in ('verified', 'published'):
-                    continue
-                # If it is assigned to a worksheet, unassign it before deletion.
-                state = workflow.getInfoFor(analysis,
-                                            'worksheetanalysis_review_state')
-                if state == 'assigned':
-                    ws = analysis.getBackReferences("WorksheetAnalysis")[0]
-                    ws.removeAnalysis(analysis)
-                # Unset the partition reference
-                analysis.edit(SamplePartition=None)
-                delete_ids.append(analysis.getId())
+
+            # Skip assigned Analyses
+            if service_uid in service_uids:
+                continue
+
+            # Skip Analyses in frozen states
+            if self._is_frozen(analysis):
```

Comment:
`_is_frozen` function is only used once, here. As per a previous comment, `cancelled` and `inactive` analyses should be considered as "frozen" states too. I'd replace this condition by the following:
```python
if not isActive(instance) or wasTransitionPerformed(instance, 'verify'):
```
Or make this the logic inside `_is_frozen` and use it in L153 too